packages/core/src/lib/markdown-it/index.js

```diff
@@ -86,43 +101,104 @@ markdownIt.renderer.rules.fence = (tokens, idx, options, env, slf) => {
-    function parseAndZeroBaseLineNumber(numberString) {
-      // authors provide line numbers to highlight based on the 'start-from' attribute if it exists
-      // so we need to shift them all back down to start at 0
-      return parseInt(numberString, 10) - startFromZeroBased;
+    function parseRule(ruleString) {
+      // Note: authors provide line numbers based on the 'start-from' attribute if it exists,
+      //       so we need to shift line numbers back down to start at 0
+
+      let ruleComponents = ruleString.split('-').map(comp => {
+        // tries to match to the line slice pattern
+        const matches = comp.match(LINESLICE_REGEX);
+        if (matches) {
+          // keep the capturing group matches only
+          let numbers = matches.slice(1);
+
+          // only the first number is a line number, the rest are regular numbers
+          numbers = numbers.map(x => x !== '' ? parseInt(x, 10) : -1);
+          numbers[0] -= startFromZeroBased;
+          return numbers;
+        }
+
+        // match fails, so it is just line numbers
+        return parseInt(comp, 10) - startFromZeroBased;
+      });
+
+      // If the only component is a line-slice, then the outer array is unnecessary as the component itself
+      // is already an array
+      const firstComponent = ruleComponents[0];
+      return ruleComponents.length === 1 && isLineSlice(firstComponent) ? firstComponent : ruleComponents;
     }
-    lineNumbersAndRanges = highlightLines.map(elem => elem.split('-').map(parseAndZeroBaseLineNumber));
+    highlightRules = highlightLines.map(parseRule);
   }
-  
+
   lines.pop(); // last line is always a single '\n' newline, so we remove it
   // wrap all lines with <span> so we can number them
   str = lines.map((line, index) => {
     const currentLineNumber = index + 1;
-    // check if there is at least one range or line number that matches the current line number
-    // Note: The algorithm is based off markdown-it-highlight-lines (https://github.com/egoist/markdown-it-highlight-lines/blob/master/src/index.js) 
+    // check if there is a highlight rule that is applicable to the line number, and handle accordingly
+    // Note: The algorithm is based off markdown-it-highlight-lines (https://github.com/egoist/markdown-it-highlight-lines/blob/master/src/index.js)
     //       This is an O(n^2) solution wrt to the number of lines
     //       I opt to use this approach because it's simple, and it is unlikely that the number of elements in `lineNumbersAndRanges` will be large
     //       There is possible room for improvement for a more efficient algo that is O(n).
-    const inRange = lineNumbersAndRanges.some(([start, end]) => {
-      if (start && end) {
-        return currentLineNumber >= start && currentLineNumber <= end;
+    // Edit (28/8/2020): I changed the approach from using some() to a simple for-loop. It is still O(n^2).
```

Comment:
let's just remove the long comment from `Note: The algorithm ...`

this area is hardly a bottleneck in performance

Reply:
Ah, okay, will do